src/games/strategy/ui/SwingWorkerCompletionWaiter.java

```diff
@@ -0,0 +1,82 @@
+package games.strategy.ui;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.awt.Window;
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+
+import javax.swing.SwingWorker;
+
+/**
+ * Manages the display of a window during the execution of a {@code SwingWorker}.
+ *
+ * <p>
+ * Use this class when it is necessary to display progress during a potentially long-running operation. If the <i>Event
+ * Dispatch Thread</i> must be blocked until the operation is complete (i.e. the user should not be allowed to perform
+ * any other actions in the UI), it is recommended you use a modal dialog.
+ * </p>
+ *
+ * <p>
+ * For example:
+ * </p>
+ *
+ * <pre>
+ * JDialog dialog = new JDialog(owner, true);
+ * swingWorker.addPropertyChangeListener(
+ *     new SwingWorkerCompletionWaiter(dialog));
+ * swingWorker.execute();
+ * // the dialog will be visible until the SwingWorker is done
+ * dialog.setVisible(true);
+ * </pre>
+ */
+public final class SwingWorkerCompletionWaiter implements PropertyChangeListener {
+  static final String SWING_WORKER_STATE_PROPERTY_NAME = "state";
+
+  private final IWindow window;
+
+  /**
+   * Initializes a new instance of the {@code SwingWorkerCompletionWaiter} class.
+   *
+   * @param window The window to display while the operation is in progress; must not be {@code null}.
+   */
+  public SwingWorkerCompletionWaiter(final Window window) {
+    this.window = IWindow.fromAwt(checkNotNull(window));
+  }
+
+  SwingWorkerCompletionWaiter(final IWindow window) {
```

Comment:
Please mark with `@VisibleForTesting`. It appears this would be private if it were not for tests.